sky/packages/sky/lib/rendering/object.dart

```diff
@@ -485,33 +618,38 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
       sky.tracing.end('RenderObject.flushPaint');
     }
   }
-
+  void initialPaint(ContainerLayer rootLayer, Size size) {
+    assert(attached);
+    assert(parent is! RenderObject);
+    assert(!_debugDoingPaint);
+    assert(hasLayer);
+    PaintingContext newContext = new PaintingContext.withLayer(rootLayer, Point.origin & size);
+    _paintLayer(newContext);
+  }
   void _repaint() {
-    assert(!_needsLayout);
-    assert(requiresCompositing);
+    assert(hasLayer);
     assert(_layer != null);
-    // TODO(abarth): Using _layer.offset isn't correct if the topLeft of our
-    // paint bounds has changed since our last repaint.
-    PaintingContext context = new PaintingContext(_layer.offset, paintBounds.size);
-    _layer.parent.add(context.layer, before: _layer);
-    _layer.detach();
-    _layer = context._layer;
-    _needsPaint = false;
+    PaintingContext newContext = new PaintingContext.replacingLayer(_layer, paintBounds);
```

Comment:
I wonder if replacingLayer is the best name.  I expected it to call _layer.replaceWith(newContext._containerLayer internally.